demos/digest/EVP_MD_demo.c

```diff
@@ -0,0 +1,189 @@
+
+int demonstrate_digest(void) {
+	OSSL_LIB_CTX	*library_context;
+	int result = 0;
+
+	library_context = OSSL_LIB_CTX_new();
+	if (library_context == NULL) {
+		fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+		ERR_print_errors_fp(stderr);
+		return 0;
+	}
+
+	/*
+	 * Fetch a message digest by name
+	 */
+	const char * digest_name = "SHA3-512";
+	const char * option_properties = NULL;
+	EVP_MD *message_digest = EVP_MD_fetch(library_context,
+								digest_name, option_properties);
+	if (message_digest == NULL) {
+		fprintf(stderr, "EVP_MD_fetch could not find %s.", digest_name);
+        ERR_print_errors_fp(stderr);
+		OSSL_LIB_CTX_free(library_context);
+		return 0;
+	}
+/*
+ * Determine the length of the fetched digest type
+ */
+	unsigned int digest_length = EVP_MD_size(message_digest);
+	if (digest_length <= 0) {
+		fprintf(stderr, "EVP_MD_size returned invalid size.\n");
+		goto cleanup_md;
+	}
+
+	unsigned char *digest_value = malloc(digest_length);
+	if (digest_value == NULL) {
+		fprintf(stderr, "No memory.\n");
+		goto cleanup_md;
+	}
+/*
+ * Make a message digest context to hold temporary state
+ * during digest creation
+ */
+	EVP_MD_CTX *digest_context = EVP_MD_CTX_new();
+	if (digest_context == NULL) {
+		fprintf(stderr, "EVP_MD_CTX_new failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md;
+	}
+/*
+ * Initialize the message digest context to use the fetched 
+ * digest provider
+ */
+	if (EVP_DigestInit(digest_context, message_digest) != 1) {
+		fprintf(stderr, "EVP_DigestInit failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+/*
+ * Digest parts one and two of the soliloqy
+ */
+	if (EVP_DigestUpdate(digest_context, hamlet_1, strlen(hamlet_1)) != 1) {
+		fprintf(stderr, "EVP_DigestUpdate(hamlet_1) failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	if (EVP_DigestUpdate(digest_context, hamlet_2, strlen(hamlet_2)) != 1) {
+		fprintf(stderr, "EVP_DigestUpdate(hamlet_2) failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	if (EVP_DigestFinal(digest_context, digest_value, &digest_length) != 1) {
+		fprintf(stderr, "EVP_DigestFinal() failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	for( int ii=0; ii<digest_length; ii++ ) {
+		fprintf(stdout, "%2.2x", digest_value[ii]);
+	}
+	fprintf(stdout, "\n");
+/*
+ * Check digest_value against the known answer
+ */
+	if ((size_t)digest_length != sizeof(known_answer)) {
+		fprintf(stdout, "Digest length(%d) not equal to known answer length(%lu).\n",
+			digest_length, sizeof(known_answer));
+	} else if (memcmp(digest_value, known_answer, digest_length) != 0) {
+		for( int ii=0; ii<sizeof(known_answer); ii++ ) {
+			fprintf(stdout, "%2.2x", known_answer[ii] );
+		}
+		fprintf(stdout, "\nDigest does not match known answer\n");
+	} else {
+		fprintf(stdout, "Digest computed properly.\n");
+		result = 1;
+	}
+
+cleanup_md_ctx:
+	EVP_MD_CTX_free(digest_context);
```

Comment:
If you init this variable to NULL, then you don't need two goto targets since all OpenSSL free routines accept NULL.